File: c++/2667.cpp
#include <iostream>
#include <queue>
#include <algorithm>
#include <vector>

using namespace std;

int N;
int map[26][26];
int visit[26][26];

int move_X[4] = {1, 0, -1, 0};
int move_Y[4] = {0, 1, 0, -1};

queue<pair<int,int>> q;
vector<int> answer;

int bfs(int x, int y){
    int answer = 0;
    visit[x][y]=1;
    q.push({x,y});
    while(!q.empty()){
        int current_x=q.front().first;
        int current_y=q.front().second;
        answer += 1;
        q.pop();
        for(int i = 0 ; i < 4 ; i++){
            int next_x = current_x + move_X[i];
            int next_y = current_y + move_Y[i];
            if(next_x>0 && next_x<=N && next_y>0 && next_y<=N){
                if(map[next_x][next_y]!=0 && visit[next_x][next_y]==0){
                    q.push({next_x,next_y});
                    visit[next_x][next_y]=1;
                }
            }
        }
    }
    return answer;
}

int main(){
    cin >> N;
    for(int i=1 ; i<=N ; i++){
        for(int j=1 ; j<=N ; j++){
            scanf("%1d",&map[i][j]);
        }
    }

    for(int i=1 ; i<=N ; i++){
        for(int j=1 ; j<=N ; j++){
            if(map[i][j]==1 && visit[i][j]==0){
                answer.push_back(bfs(i,j));
            }
        }
    }

    sort(answer.begin(),answer.end());
    cout << answer.size() << "\n";
    for(int i=0 ; i<answer.size() ; i++){
        cout << answer[i] << "\n";
    }
    return 0;
}

File: python/2667.py
N = int(input())
miro = []
for i in range(N):
    miro.append(list(map(int, input())))

finder = [[0 for i in range(N)] for j in range(N)]

move_x = [0, 1, 0, -1]
move_y = [1, 0, -1, 0]

answer = []


def bfs(x, y):
    queue = []
    queue.insert(0, [x, y])
    finder[x][y] = 1
    temp = 1
    while len(queue) > 0 :
        now_x = queue[-1][0]
        now_y = queue[-1][1]
        queue.pop()
        for idx in range(4):
            next_x = now_x + move_x[idx]
            next_y = now_y + move_y[idx]
            if (next_x >= 0) and (next_y >= 0) and (next_x <= N-1) and (next_y <= N-1):
                if miro[next_x][next_y] == 1 and finder[next_x][next_y] == 0:
                    queue.insert(0, [next_x, next_y])
                    finder[next_x][next_y] = 1
                    temp += 1

    return temp


for i in range(N):
    for j in range(N):
        if finder[i][j] == 0 and miro[i][j] == 1:
            answer.append(bfs(i, j))

print(len(answer))
answer = sorted(answer)
for i in answer:
    print(i)
